### PR TITLE
Webapp Slots

### PR DESCRIPTION
**Pull request by LukaszStem**
Add: Slot support for all webapp commands
Add: Tests for commands with slots
Fix: Minor description fixes
Edit: Minor optimizations for appsettings and hostnames

### CODE REVIEW

File: lib/commands/arm/webapp/webapp._js [hunk 1]
```diff
@@ -89,26 +89,26 @@ exports.init = function (cli) {
...
       }
+      var subscription = profile.current.getSubscription(options.subscription);
+      var client = webappUtils.createWebappManagementClient(subscription);
       var progress = cli.interaction.progress(util.format($('Stopping webapp %s'), name));
-      var result;
       try {
-        result = client.sites.stopSite(resourceGroup, name, options, _);
+        if (Boolean(options.slot)) {
+          client.sites.stopSiteSlot(resourceGroup, name, options.slot, _);
+          log.info('Webapp slot' + name + '/' + options.slot + ' has been stopped ');
```
Comment (yugangw-msft) [inline lib/commands/arm/webapp/webapp._js]: need a space after `Webapp slot`. Consider using `util.format`

File: lib/commands/arm/webapp/webapp._js [hunk 2]
```diff
@@ -62,23 +60,25 @@ exports.init = function (cli) {
...
 	    var siteEnvelope = {
-	       sitename: webSiteSlotName,
+	       sitename: name,
 		     serverFarmId: plan,
 		     location: location
 	    };
-      var result;
       try {
-        result = client.sites.createOrUpdateSite(resourceGroup, name, siteEnvelope, _);
+        if (Boolean(options.slot)) {
```
Comment (yugangw-msft) [inline lib/commands/arm/webapp/webapp._js]: you can just use `if (options.slot)`, empty string is a false value

File: lib/commands/arm/webapp/webapp._js [hunk 3]
```diff
@@ -230,21 +243,48 @@ exports.init = function (cli) {
...
-          log.table(data, function (row, item) {
-		    var serverFarmIdArr = (item.serverFarmId).split('/');
-		    var serverFarmId = serverFarmIdArr[serverFarmIdArr.length - 1];
-            row.cell($('Name '), item.name);
-            row.cell($('Status '), item.state);
-            row.cell($('Location '), item.location);
-            row.cell($('App Service Plan '), serverFarmId);
-            row.cell($('URL '), item.hostNames);
+
+      if (result !== null && result !== undefined && result.length > 0) {
```
Comment (yugangw-msft) [inline lib/commands/arm/webapp/webapp._js]: can be simplified `result && result.length > 0`

File: lib/commands/arm/webapp/webapp.config._js [hunk 4]
```diff
@@ -659,15 +719,21 @@ exports.init = function (cli) {
...
+        } else {
+          currentAppSettings = client.sites.listSiteAppSettings(resourceGroup, name, _);
+        }
         delete currentAppSettings.properties['DOCKER_REGISTRY_SERVER_URL'];
         delete currentAppSettings.properties['DOCKER_REGISTRY_SERVER_USERNAME'];
         delete currentAppSettings.properties['DOCKER_REGISTRY_SERVER_PASSWORD'];
         delete currentAppSettings.properties['DOCKER_CUSTOM_IMAGE_NAME'];
-        currentAppSettings.location = result.location;
-        result = client.sites.updateSiteAppSettings(resourceGroup, name, currentAppSettings, _);
+        progress = cli.interaction.progress($('Deleting webapp container settings'));
```
Comment (yugangw-msft) [inline lib/commands/arm/webapp/webapp.config._js]: Progress of `Getting webapp configuration` is already going (see about 10 lines above), so it is not correct to start another one. Suggest remove this, and update the 1st progress text be `Deleting webapp container settings`

File: lib/commands/arm/webapp/webapp._js (hunk 1 above)
Comment (LukaszStem) [inline lib/commands/arm/webapp/webapp._js]: fixed the space

File: lib/commands/arm/webapp/webapp._js (hunk 2 above)
Comment (LukaszStem) [inline lib/commands/arm/webapp/webapp._js]: fixed

File: lib/commands/arm/webapp/webapp._js (hunk 3 above)
Comment (LukaszStem) [inline lib/commands/arm/webapp/webapp._js]: fixed

File: lib/commands/arm/webapp/webapp.config._js (hunk 4 above)
Comment (LukaszStem) [inline lib/commands/arm/webapp/webapp.config._js]: fixed, updated the 1st progress text

File: lib/commands/arm/webapp/webapp.config._js [hunk 5]
```diff
@@ -574,24 +623,31 @@ exports.init = function (cli) {
...
         if (options.password) 
           currentAppSettings.properties['DOCKER_REGISTRY_SERVER_PASSWORD'] = options.password;
-        if (options.customimagename) 
-          currentAppSettings.properties['DOCKER_CUSTOM_IMAGE_NAME'] = options.customimagename;
-        result = client.sites.getSite(resourceGroup, name, _);
-        currentAppSettings.location = result.location;
-        result = client.sites.updateSiteAppSettings(resourceGroup, name, currentAppSettings, _);
+        if (options['customImageName'])
+          currentAppSettings.properties['DOCKER_CUSTOM_IMAGE_NAME'] = options['customImageName'];
+        progress = cli.interaction.progress($('Setting webapp container settings'));
```
Comment (yugangw-msft) [inline lib/commands/arm/webapp/webapp.config._js]: again, the `progress` conflicts with `Getting webapp container settings`

File: test/commands/arm/webapp/arm.webapp.feature-tests.js [hunk 6]
```diff
@@ -116,90 +122,190 @@ describe('arm', function () {
...
     it('config set should work', function (done) {
       suite.execute('webapp config set %s %s --phpversion %s --json', groupName, sitename, updatedPHPValue, function (result) {
         result.exitStatus.should.equal(0);
         done();
       });
     });
 
+    it('config slot set should work', function (done) {
+      suite.execute('webapp config set %s %s --slot %s --phpversion %s  --json', groupName, sitename, slotName, updatedPHPValue, function (result) {
+        result.exitStatus.should.equal(0);
```
Comment (yugangw-msft) [inline test/commands/arm/webapp/arm.webapp.feature-tests.js]: Should you do a little more verification that your php version change did go through? Just add one more line below verify the the `phpVersion` from the `result`

Comment (LukaszStem) [inline test/commands/arm/webapp/arm.webapp.feature-tests.js]: fixed

File: test/commands/arm/webapp/arm.webapp.feature-tests.js [hunk 7]
```diff
@@ -137,69 +168,146 @@ describe('arm', function () {
       });
     });
 
+    it('config slot appsettings set should work', function (done) {
+      suite.execute('webapp config appsettings set %s %s %s --slot %s --json', groupName, sitename, finalAppSetting, slotName, function (result) {
+        result.exitStatus.should.equal(0);
```
Comment (yugangw-msft) [inline test/commands/arm/webapp/arm.webapp.feature-tests.js]: Like we agreed, please do not forget to add detail test in your next PR

**Pull request merged**